pkg/kubelet/util/pod_startup_latency_tracker.go

```diff
@@ -0,0 +1,186 @@
+	// log, if pod latency was already Observed
+	metricRecorded bool
+}
+
+// default
+func NewPodStartupLatencyTracker() PodStartupLatencyTracker {
+	return &basicPodStartupLatencyTracker{
+		pods:  map[types.UID]*perPodState{},
+		clock: clock.RealClock{},
+	}
+}
+
+func (p *basicPodStartupLatencyTracker) ObservedPodOnWatch(pod *v1.Pod, when time.Time) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// if the pod is terminal, we do not have to track it anymore for startup
+	if pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodSucceeded {
+		delete(p.pods, pod.UID)
+		return
+	}
+
+	state := p.pods[pod.UID]
+	if state == nil {
+		// create a new record for pod, only if it was not yet acknowledged by the Kubelet
+		// this is required, as we want to log metric only for those pods, that where scheduled
+		// after Kubelet started
+		if pod.Status.StartTime.IsZero() {
+			p.pods[pod.UID] = &perPodState{}
+		}
+
+		return
+	}
+
+	if state.observedRunningTime.IsZero() {
+		// skip, pod didn't start yet
+		return
+	}
+
+	if state.metricRecorded {
+		// skip, pod's latency already recorded
+		return
+	}
+
+	if hasPodStartedSLO(pod) {
+		podStartingDuration := when.Sub(pod.CreationTimestamp.Time)
+		imagePullingDuration := state.lastFinishedPulling.Sub(state.firstStartedPulling)
+		podStartSLOduration := (podStartingDuration - imagePullingDuration).Seconds()
+
+		klog.InfoS("Observed pod startup duration",
+			"pod", klog.KObj(pod),
+			"podStartSLOduration", podStartSLOduration,
+			"pod.CreationTimestamp", pod.CreationTimestamp.Time,
+			"firstStartedPulling", state.firstStartedPulling,
+			"lastFinishedPulling", state.lastFinishedPulling,
+			"observedRunningTime", state.observedRunningTime,
+			"watchObservedRunningTime", when)
+
+		metrics.PodStartSLIDuration.WithLabelValues().Observe(podStartSLOduration)
+		state.metricRecorded = true
+	}
+}
+
+func (p *basicPodStartupLatencyTracker) RecordImageStartedPulling(podUID types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[podUID]
+	if state == nil {
+		return
+	}
+
+	if state.firstStartedPulling.IsZero() {
+		state.firstStartedPulling = p.clock.Now()
+	}
+}
+
+func (p *basicPodStartupLatencyTracker) RecordImageFinishedPulling(podUID types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[podUID]
+	if state == nil {
+		return
+	}
+
+	state.lastFinishedPulling = p.clock.Now() // Now is always grater than values from the past.
+}
+
+func (p *basicPodStartupLatencyTracker) RecordStatusUpdated(pod *v1.Pod) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[pod.UID]
+	if state == nil {
+		return
+	}
+
+	if state.metricRecorded {
+		// skip, pod latency already recorded
+		return
+	}
+
+	if !state.observedRunningTime.IsZero() {
+		// skip, pod already started
+		return
+	}
+
+	if hasPodStartedSLO(pod) {
+		klog.V(2).InfoS("Mark when the pod was running for the first time", "pod", klog.KObj(pod), "rv", pod.ResourceVersion)
```

Comment:
Does this need to be v=2 verbosity? v=2 is default verbosity for many distros and we already get logs about time when pod is running. Maybe increase this if it is for debugging purposes?

Reply:
Agreed, normally this would be a 4